performance_test/src/communication_abstractions/resource_manager.cpp

```diff
@@ -239,4 +244,78 @@ void ResourceManager::connext_dds_micro_subscriber(
   }
 }
 #endif
+
+#ifdef PERFORMANCE_TEST_OPENDDS_ENABLED
+DDS::DomainParticipant_ptr
+ResourceManager::opendds_participant() const
+{
+  std::lock_guard<std::mutex> lock(m_global_mutex);
+  
+  if (CORBA::is_nil(m_opendds_participant)) {
+   DDS::DomainParticipantFactory_var dpf = TheParticipantFactory;
+
+    TransportConfig_rch config = TransportRegistry::instance()->create_config("ApexAiConfig");
+    TransportInst_rch inst = TransportRegistry::instance()->create_inst("rtps_tran","rtps_udp");
+ 
+    RtpsUdpInst_rch rui = static_rchandle_cast<RtpsUdpInst>(inst);
+    rui->handshake_timeout_ = 1;
+
+    config->instances_.push_back(inst);
+    TransportRegistry::instance()->global_config(config);
+
+    int domain = 0;
+    bool multicast = true;
+    unsigned int resend = 1;
+    std::string partition, governance, permissions;
+    int defaultSize = 0;
+
+    RtpsDiscovery_rch disc = make_rch<RtpsDiscovery>("RtpsDiscovery");
+    rui->use_multicast_ = true;
+    rui->local_address("127.0.0.1:");
+    rui->multicast_interface_="lo";
+    disc->sedp_multicast(true);
+
+    TheServiceParticipant->add_discovery(static_rchandle_cast<Discovery>(disc));
+    TheServiceParticipant->set_repo_domain(domain, disc->key());
+    DDS::DomainParticipantQos dp_qos;
+    dpf->get_default_participant_qos(dp_qos);
+    m_opendds_participant = dpf->create_participant(m_ec.dds_domain_id(),PARTICIPANT_QOS_DEFAULT,0,DEFAULT_STATUS_MASK);
+  }
+  return m_opendds_participant;
+}
+
+void
+ResourceManager::opendds_publisher(DDS::Publisher_ptr & publisher, DDS::DataWriterQos & dw_qos) const
+{
+  DDS::DomainParticipant_ptr participant = opendds_participant();
+  std::lock_guard<std::mutex> lock(m_global_mutex);
+
+  publisher = participant->create_publisher(PUBLISHER_QOS_DEFAULT,0,OpenDDS::DCPS::DEFAULT_STATUS_MASK);
+
+  if (CORBA::is_nil(publisher))
+     throw std::runtime_error("Failed to create publisher");
+
+  DDS::ReturnCode_t ret;
+  ret = publisher->get_default_datawriter_qos(dw_qos);
+  if (ret != DDS::RETCODE_OK)
+    throw std::runtime_error("Failed to get default datawriter qos");
+}
+
+void
+ResourceManager::opendds_subscriber(DDS::Subscriber_ptr & subscriber, DDS::DataReaderQos & dr_qos) const
+{
+  DDS::DomainParticipant_ptr participant = opendds_participant();
+  std::lock_guard<std::mutex> lock(m_global_mutex);
+
+  subscriber = participant->create_subscriber(SUBSCRIBER_QOS_DEFAULT,0,OpenDDS::DCPS::DEFAULT_STATUS_MASK);
```

Comment:
Can you use `nullptr` for the listener pointer?

Reply:
done